### PR TITLE
RN-156: volume management function

### PR DESCRIPTION
**Pull request by bryan-huangyan**

### CODE REVIEW

File: web/src/routes/volume.go [hunk 1]
```diff
@@ -111,18 +204,62 @@ func (a *VolumeAdmin) Delete(ctx context.Context, id int64) (err error) {
...
+	db := DB()
+	db = db.Begin()
+	defer func() {
+		if err == nil {
+			db.Commit()
+		} else {
+			db.Rollback()
+		}
+	}()
+	volume := &model.Volume{}
```
Comment (Catherine2019) [inline web/src/routes/volume.go:240]: Need to check owner of the volume

File: web/src/routes/volume.go [hunk 2]
```diff
@@ -29,17 +30,92 @@ var (
...
+		log.Println("Not authorized to read the volume")
+		err = fmt.Errorf("Not authorized")
+		return
+	}
+	return
+}
+
+func (a *VolumeAdmin) Create(ctx context.Context, name string, size int32,
+	iopsLimit int32, iopsBurst int32, bpsLimit int32, bpsBurst int32, poolID string) (volume *model.Volume, err error) {
+	memberShip := GetMemberShip(ctx)
```
Comment (Catherine2019) [inline web/src/routes/volume.go:81]: Need to check permission

File: scripts/kvm/detach_volume_wds_vhost.sh [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    echo "|:-COMMAND-:| $(basename $0) '' '$vol_ID'"
+fi
+vm_xml=$xml_dir/$vm_ID/$vm_ID.xml
+virsh dumpxml --security-info $vm_ID 2>/dev/null | sed "s/autoport='yes'/autoport='no'/g" > $vm_xml.dump && mv -f $vm_xml.dump $vm_xml
+
+vhost_name=instance-$vm_ID-vol-$vol_ID
+vhost_id=$(wds_curl GET "api/v2/sync/block/vhost" | jq --arg vhost $vhost_name -r '.vhosts | .[] | select(.name == $vhost) | .id')
+uss_id=$(wds_curl GET "api/v2/wds/uss" | jq --arg hname $(hostname -s) -r '.uss_gateways | .[] | select(.server_name == $hname) | .id')
+wds_curl PUT "api/v2/sync/block/vhost/unbind_uss" "{\"vhost_id\": \"$vhost_id\", \"uss_gw_id\": \"$uss_id\", \"is_snapshot\": false}"
+wds_curl DELETE "api/v2/sync/block/vhost/$vhost_id"
```
Comment (Catherine2019) [inline scripts/kvm/detach_volume_wds_vhost.sh:25]: Better to add a new line

File: scripts/xml/wds_volume.xml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<disk type='vhostuser' device='disk'>
+  <driver name='qemu' type='raw'/>
+  <source type='unix' path='VM_UNIX_SOCK'/>
+  <target dev='VOLUME_TARGET' bus='virtio'/>
+</disk>
```
Comment (Catherine2019) [inline scripts/xml/wds_volume.xml:5]: Better to add a new line

File: web/src/apis/volume.go [hunk 5]
```diff
@@ -21,11 +25,41 @@ var volumeAdmin = &routes.VolumeAdmin{}
...
+	IopsLimit int32  `json:"iops_limit" binding:"omitempty,gte=0"`
+	IopsBurst int32  `json:"iops_burst" binding:"omitempty,gte=0"`
+	BpsLimit  int32  `json:"bps_limit" binding:"omitempty,gte=0"`
+	BpsBurst  int32  `json:"bps_burst" binding:"omitempty,gte=0"`
+}
+
+type VolumePatchPayload struct {
+	Name       string `json:"name" binding:"omitempty"`
+	Size       int32  `json:"size" binding:"omitempty"`
+	InstanceID string `json:"instance_id" binding:"omitempty"`
```
Comment (Catherine2019) [inline web/src/apis/volume.go:42]: By convention, InstanceID can be defined as Instance *BaseID

File: scripts/xml/wds_volume.xml (hunk 4 above)
Comment (bryan-huangyan) [reply scripts/xml/wds_volume.xml]: fixed

File: scripts/kvm/detach_volume_wds_vhost.sh (hunk 3 above)
Comment (bryan-huangyan) [reply scripts/kvm/detach_volume_wds_vhost.sh]: fixed

File: web/src/routes/volume.go (hunk 2 above)
Comment (bryan-huangyan) [reply web/src/routes/volume.go:81]: fixed

File: web/src/routes/volume.go (hunk 1 above)
Comment (bryan-huangyan) [reply web/src/routes/volume.go:258]: fixed

**Review by Catherine2019 — APPROVED** [review]

**Pull request merged**